cpp_faster_fifo/cpp_lib/faster_fifo.cpp

```diff
@@ -145,28 +145,42 @@ bool timer_positive(const struct timeval &timer) {
     return (timer.tv_sec > 0) || (timer.tv_sec == 0 && timer.tv_usec > 0);
 }
 
-int queue_put(void *queue_obj, void *buffer, void *msg_data, size_t msg_size, int block, float timeout) {
+int queue_put(void *queue_obj, void *buffer, const void **msgs_data, const size_t *msg_sizes, const size_t num_msgs, const int block, const float timeout) {
     auto q = (Queue *)queue_obj;
     LockGuard lock(&q->mutex);
 
     auto wait_remaining = float_seconds_to_timeval(timeout);
-    while (!q->can_fit(msg_size + sizeof(msg_size))) {
-        if (!block || !timer_positive(wait_remaining))
-            return Q_FULL;
 
-        wait_remaining = wait(wait_remaining, &q->not_full, &q->mutex);
-    }
+    bool has_written = false;
+    for (size_t i = 0; i < num_msgs; ++i) {
+        const size_t next_msg_size = msg_sizes[i];
+        while (!q->can_fit(next_msg_size + sizeof(size_t))) {
+            if (has_written)
+                pthread_cond_signal(&q->not_empty);
+
+            has_written = false;
+
+            if (!block || !timer_positive(wait_remaining))
+                return Q_FULL;
```

Comment:
I think my main problem with this is the fact that the function can successfully write a few messages to the Queue, and then return Q_FULL.

Q_FULL can be interpreted as an error by the user code, and since now there is no way to determine how many message were added to the queue, it's impossible to recover from this correctly.

Now I remember this was one of the reasons I decided against implementing this function, because this really complicates things.

Reply:
Note that there isn't a similar situation in get_many.

We return EMPTY if and only if we failed to read even a single message.

If there are more messages that we can read into our memory buffer, we return those that fit, and then reallocate the buffer to fit more messages in the future.

Reply:
I see value in put_many though.
It is not the same as q.put([msg1, msg2...]) because you need no extra code on the receiving end.

Do you think we can reimplement `put` to address my concerns above?
So I see two options:
1. The function return SUCCESS only if all messages fit, and FULL otherwise. This is hacky and will fail in some cases (i.e. when there was enough space to fit half of the messages, but we return FULL instead). But I like the simplicity of this. And if users need more sophisticated behavior they can always do their own put_many loop, which will only be marginally slower.
 This idea also has an additional benefit of having to signal cond variable only once
 
2. Alternatively, we can change the interface to return the number of messages which were actually sent. And have the user handle re-sending of the rest of the messages. This is more flexible, but honestly I prefer option 1.

Reply:
Went with 1 here.  That's the simplest and it keeps faster_fifo working more similar to a normal Queue